File: impls/rust1_94_0/src/bin/step1_read_print.rs
use rust1_94_0::{
    error::MalError,
    printer::pr_str,
    reader::read_str,
    readline as rl,
    types::{MalObject, MalResult},
};

fn read(input: &str) -> MalResult {
    read_str(input)
}

fn eval(ast: &MalObject) -> MalObject {
    ast.clone()
}

fn print(ast: &MalObject) -> String {
    pr_str(ast, true)
}

fn rep(input: &str) -> Result<String, MalError> {
    Ok(print(&eval(&read(input)?)))
}

fn main() {
    loop {
        let input = rl::readline("user> ").unwrap();
        let result = rep(&input);
        match result {
            Ok(output) => println!("{}", output),
            Err(MalError::ParseError(message)) => println!("{}", message),
            _ => {}
        }
    }
}

File: impls/ocaml541/bin/step1_read_print.ml
open Ocaml541.Rl
open Ocaml541.Reader
open Ocaml541.Printer
open Result.Syntax

let read str = read_str str
let eval ast = Ok ast
let print ast = pr_str ast true

let rep str =
  let* ast = read str in
  let* exp = eval ast in
  Ok (print exp)

let () =
  let rec loop () =
    let result = readline ~prompt:"user> " () in
    match result with
    | Some line ->
        (match rep line with
        | Ok out -> print_endline out
        | Error err ->
            let err_msg = Format.asprintf "Error: %s" (pr_str err true) in
            print_endline err_msg);
        loop ()
    | None -> loop ()
  in
  loop ()
